app/Libraries/Chat.php

```diff
@@ -28,6 +30,48 @@ public static function ack(User $user)
         $transaction->exec();
     }
 
+    public static function createAnnouncement(User $sender, array $rawParams)
+    {
+        priv_check_user($sender, 'ChatAnnounce')->ensureCan();
+
+        $params = get_params($rawParams, null, [
+            'channel:any',
+            'message:string',
+            'target_ids:int[]',
+        ], ['null_missing' => true]);
+
+        if (!isset($params['target_ids'])) {
+            throw new InvariantException('missing target_ids parameter');
+        }
+
+        if (!isset($rawParams['channel'])) {
+            throw new InvariantException('missing channel parameter');
+        }
+
+        $users = User::whereIn('user_id', $params['target_ids'])->get();
+        if ($users->isEmpty()) {
+            throw new InvariantException('Nobody to broadcast to!');
+        }
+
+        $users = $users->push($sender)->uniqueStrict('user_id');
+
+        $channel = (new Channel())->getConnection()->transaction(function () use ($sender, $params, $users) {
+            $channel = Channel::createAnnouncement($users, $params['channel']);
+
+            foreach ($users as $user) {
+                event(new ChatChannelEvent($channel, $user, 'join'));
+            }
+
```

Comment:
also applies for sending pm but failing sending the actual message result in empty channel on the receiver side...

Reply:
ironically because the `event`s are being called in the function instead of after the transaction completes...